### PR TITLE
Another round of various fixes

### PR DESCRIPTION
**Pull request by pulsejet**
1. [FEAT] Implement route inheritance in RIB
2. Fix PIT entry lifetimes to use Interest lifetime
3. Fix incorrect outgoing PIT token (race)
4. [FEAT] Implement suppression in best route and multicast strategy
5. Fix the (completely broken) expiry priority queue for PIT entries
6. Remove some utils in favor of standard library
7. Some other refactor

This also has some hacks I need to fix some looping. The same behavior is observable with NFD.
These are marked with `// TODO`

### CODE REVIEW

**Comment by pulsejet** [comment]
> 3. Fix incorrect outgoing PIT token (race)

I'm still not positive PIT tokens are working as ideal, so this is just YaAttempt to fix it.

File: utils/priority_queue/priority-queue.go [hunk 1]
```diff
@@ -32,13 +33,13 @@ func (pq *wrapper[V, P]) Swap(i, j int) {
...
-func (pq *wrapper[V, P]) Push(x *item[V, P]) {
-	item := x
+func (pq *wrapper[V, P]) Push(x any) {
+	item := x.(*Item[V, P])
 	item.index = len(*pq)
 	*pq = append(*pq, item)
 }
 
-func (pq *wrapper[V, P]) Pop() *item[V, P] {
+func (pq *wrapper[V, P]) Pop() any {
```
Comment (zjkmxy) [inline utils/priority_queue/priority-queue.go:42]: Why did you change it to any?

File: utils/priority_queue/priority-queue.go [hunk 2]
```diff
@@ -32,13 +33,13 @@ func (pq *wrapper[V, P]) Swap(i, j int) {
 	(*pq)[j].index = j
 }
 
-func (pq *wrapper[V, P]) Push(x *item[V, P]) {
-	item := x
+func (pq *wrapper[V, P]) Push(x any) {
```
Comment (zjkmxy) [inline utils/priority_queue/priority-queue.go:36]: Why did you change it to any? If it only supports `(*Item[V, P])` and does not expect `item` to be `nil`, I think explicit type is better.

File: utils/priority_queue/priority-queue.go (hunk 1 above)
Comment (pulsejet) [reply utils/priority_queue/priority-queue.go:42]: `container/heap` functions require this to implement this interface. It doesn't accept anything else :/

```go
type Interface interface {
	sort.Interface
	Push(x any) // add x as element Len()
	Pop() any   // remove and return element Len() - 1.
}
```

File: utils/priority_queue/priority-queue.go (hunk 2 above)
Comment (pulsejet) [reply utils/priority_queue/priority-queue.go:36]: Same as above

File: fw/thread.go [hunk 3]
```diff
@@ -430,32 +449,41 @@ func (t *Thread) processIncomingData(packet *defn.Pkt) {
...
 
 		// Insert into dead nonce list
-		for _, outRecord := range pitEntries[0].OutRecords() {
+		for _, outRecord := range pitEntry.OutRecords() {
 			t.deadNonceList.Insert(data.NameV, outRecord.LatestNonce)
 		}
 
 		// Clear out records from PIT entry
-		pitEntries[0].ClearOutRecords()
+		// TODO: NFD dev guide specifies in-records should not be cleared - why?
```
Comment (zjkmxy) [inline fw/thread.go:472]: Don't know either. I think the PIT record is completely converted into a CS record in DPDK.

File: fw/thread.go [hunk 4]
```diff
@@ -298,22 +299,38 @@ func (t *Thread) processIncomingInterest(packet *defn.Pkt) {
...
+	if fhName != nil {
+		lookupName = fhName
 	}
 
-	strategy.AfterReceiveInterest(packet, pitEntry, incomingFace.FaceID(), nexthops)
+	// Query the FIB for possible nexthops
+	nexthops := table.FibStrategyTable.FindNextHopsEnc(lookupName)
+
+	// Exclude faces that have an in-record for this interest
+	// TODO: unclear where NFD dev guide specifies such behavior (if any)
```
Comment (zjkmxy) [inline fw/thread.go:323]: I guess that's to prevent loop: an incoming face indicates that path does not have the desired Data.

**Review by zjkmxy — APPROVED** [review]

File: fw/thread.go (hunk 4 above)
Comment (pulsejet) [reply fw/thread.go:323]: Yeah, but I didn't see this condition mentioned in the dev guide .. seems kinda obvious, maybe I missed it ...

Comment (Pesa) [reply fw/thread.go:323]: FYI, the NFD dev guide is not intended as a full specification of all NDN forwarding rules, it's supposed to "just" describe the implementation in NFD. I lost count of how many times we tried (and never finished) to write a generic and implementation-agnostic spec of the NDN forwarding behavior 😞
